docs/typescript.md

```diff
@@ -2,13 +2,504 @@
+2) It is determined to be the most accurate _and_ specific type assignable.
+
+#### Acceptable usages of `:` annotations
+
+##### To prevent or fix type assertions
+
+Type annotations are more responsive to code drift than assertions. If the assignee's type becomes incompatible with the assigned type annotation, the compiler will raise a type error, whereas in most cases a type assertion will still suppress the error.
+
+When the compiler is in doubt, an annotation will nudge it towards relying on type inference, while an assertion will force it to accept the user-supplied type.
+
+<!-- TODO: Add example -->
+
+##### When instantiating an empty container type, provide a type annotation
+
+This is one case where type inference is unable to reach a useful conclusion without user-provided information. Since the compiler cannot arbitrarily restrict the range of types that could be inserted into the container, it has to assume the widest type, which is often `any`. It's up to the user to narrow that into the intended type by adding an explicit annotation.
+
+🚫
+
+```typescript
+const tokens = []; // Type 'any[]'
+const tokensMap = new Map(); // Type 'Map<any, any>'
+```
+
+✅
+
+```typescript
+const tokens: string[] = []; // Type 'string[]'
+const tokensMap = new Map<string, Token>(); // Type 'Map<string, Token>'
+```
+
+### Type Assertions
+
+`as` assertions are unsafe. They overwrite type-checked and inferred types with user-supplied types that suppress compiler errors.
+
+They should only be introduced into the code if the accurate type is unreachable through other means.
+
+##### Document safe or necessary use of type assertions
+
+When a type assertion is absolutely necessary due to constraints or is even safe due to runtime checks, we should document the reasoning behind its usage in the form of a comment.
+
+<!-- TODO: Add example -->
+
+#### Avoid `as`
+
+Type assertions make the code brittle against changes.
+
+While TypeScript and ESLint will flag some unsafe, structurally unsound, or redundant type assertions, they will generally accept user-supplied types without further type-checking.
+
+This can cause silent failures or false negatives where errors are suppressed. This is especially damaging as the codebase accumulates changes over time. Type assertions may continue to silence errors, even though the type itself, or the type's relationship to the rest of the code may have been altered so that the asserted type is no longer valid.
+
+##### Redundant or unnecessary `as` assertions are not flagged for removal
+
+Type assertions can also cause false positives, because they assertions are independent expressions, untied to the type errors they were intended to fix. Thus, even if code drift fixes or removes a particular type error, the type assertions that were put in place to fix that error will provide no indication that they are no longer necessary and now should be removed.
```

Comment:
```suggestion
Type assertions can also cause false positives, because assertions are independent expressions, untied to the type errors they were intended to fix. Thus, even if code drift fixes or removes a particular type error, the type assertions that were put in place to fix that error will provide no indication that they are no longer necessary and now should be removed.
```

Reply:
Thanks for pointing out the awkward phrasing here.

```suggestion
Type assertions can also cause false positives, because they are independent expressions, untied to the type errors they are intended to fix. Thus, even if code drift fixes or removes a particular type error, the type assertions that were put in place to fix that error will provide no indication that they are no longer necessary and should now be removed.
```